solutions/usaco-599.mdx

```diff
@@ -31,127 +31,172 @@ a one-base polynomial hash to check if the substring appears only once. If we
-				else k.push_back(1100001);
-			} else if(arr[i].s - arr[pre].s < 0)
-			{
-				if(arr[nxt].f - arr[i].f > 0) k.push_back(1100001);
-				else k.push_back(1100002);
-			}
+	vector<int> ang(n);
+	for (int i = 0; i < n; i++) {
+		int prev = (i - 1 + n) % n;
+		int nxt = (i + 1) % n;
+
+		/*
+		 * two type of possible angles (90 deg either on inside or outside)
+		 * reminder: all edges are parallel to either x or y axis & alternate
+		 *
+		 * |--------------------------|
+		 * | ^                     ^  |
+		 * |      angles of 90°       |
+		 * | ˅                     ˅  |
+		 * |--------------------------|
+		 *
+		 * |-------------------------|
+		 * |     angles of 270°      |
+		 * |              > ----------
+		 * |                |
+		 * |                |
+		 * |-----------------
+		 */
+
+		// derived from slope
+		ll rt = (dt[i].y - dt[prev].y) * (dt[nxt].x - dt[i].x) -
+						(dt[nxt].y - dt[i].y) * (dt[i].x - dt[prev].x);
+
+		// no angle should ever be colinear
+		assert(rt);
+
+		if (rt > 0) {
+			ang[i] = MX + 2;
+		} else {
+			ang[i] = MX + 1;
 		}
 	}
-	k.push_back(abs(arr[0].f-arr[n-1].f)+abs(arr[0].s-arr[n-1].s));
-	m = k.size()+1;
-	// for(int i = 0; i < m; i++) cerr << k[i] << " ";
-	cerr << endl;
-	gen_dist();
-	gen_hash();
-	dist[0] = 0;
+
+	vector<int> hsh(1);
+	hsh[0] = 0;
+
+	vector<ll> poly(2 * n + 1);
+
+	poly[0] = 1;
+	for (int i = 1; i < (int)(poly.size()); i++) {
+		poly[i] = (int)(poly[i - 1] * P % MOD);
+	}
+
+	for (int i = 1; i < n; i++) {
+		hsh.push_back((hsh.back() * P + dst[(i - 1 + n) % n][i][0]) % MOD);
+		hsh.push_back((hsh.back() * P + ang[i]) % MOD);
+	}
+	hsh.push_back(hsh.back() * P + dst[n - 1][0][0]);
+
+	function<int(int i, int j)> get_hash = [&](int i, int j) {
+		// gets the hash for the substring from i -> j
+		// you can also compute this using the mod inverse of poly_i
+
+		return ((hsh[j] - ((hsh[i] * poly[j - i]) % MOD) + MOD) % MOD);
+	};
+
 	int ans = 0;
-	for(int sp = 0; sp < m-1; sp++)
-	{
-		int know = -1;
-		for(int L = 1; L < m-sp; L++)
-		{
-			int x = (H[sp+L] - ((ll)H[sp]*p[L])%M + M)%M;
-			int nocc = 0;
-			for(int pt = 0; pt < m-L; pt++)
-			{
-				if((H[pt+L] - ((ll)H[pt]*p[L])%M + M)%M == x)
-					nocc++;
+
+	for (int i = 1; i < (int)(hsh.size()) - 2; i++) {
+		int fin = -1;
+
+		for (int len = 1; len < (int)(hsh.size()) - i; len++) {
+			int cr = get_hash(i, i + len);
+			int occ = 0;
+
+			for (int j = 0; j < (int)(hsh.size()) - len; j++) {
+				if (get_hash(j, j + len) == cr) {
+					occ++;
+				}
 			}
-			if(nocc == 1)
-			{
-				know = (sp+L+1)/2;
+
+			// bessie can uniquely identify her position
+			if (occ == 1) {
+				fin = (i + len + 1) / 2;
 				break;
 			}
 		}
-		int a = cwdist[(sp+2)/2][0];
-		if(know != -1)
-		{
-			a = dist[know] + cwdist[(sp+2)/2][know];
+
+		int strt = (i + 2) / 2;
+		if (fin >= n) {
+			fin = -1;
 		}
-		int b = dist[(sp+2)/2];
-		// cerr << a << " " << b << endl;
-		ans = max(ans, a-b);
+
+		int a = dst[strt][0][0];
```

Comment:
please don't remove vowels from the var names

Reply:
☃️